Lib/modulefinder.py

```diff
@@ -43,6 +51,67 @@ def ReplacePackage(oldname, newname):
     replacePackageMap[oldname] = newname
 
 
+def _find_module(name, path=None):
+    """An importlib reimplementation of imp.find_module (for our purposes)."""
+
+    importlib.machinery.PathFinder.invalidate_caches()
+
+    spec = importlib.machinery.PathFinder.find_spec(name, path+sys.path)
+
+    if spec is None:
+        raise ImportError("No module named {name!r}".format(name=name), name=name)
+
+    if isinstance(spec.loader, type):  # Some special cases:
+
+        if issubclass(spec.loader, importlib.machinery.BuiltinImporter):
+            return None, None, ("", "", _C_BUILTIN)
+
+        if issubclass(spec.loader, importlib.machinery.FrozenImporter):
+            return None, None, ("", "", _PY_FROZEN)
+
+    file_path = spec.origin
+
+    try:
+
+        # `imp` behavior is s bit funny here:
+
+        # If the path is a descendant of the CWD, it is reported as a relative filepath
+        # ...unless it is an extension, in which case it will be absolute (see below)
+        # ...otherwise, it is fully resolved.
+
+        if os.path.commonpath(map(os.path.abspath, (file_path, os.getcwd()))) == os.getcwd():
+            file_path = os.path.relpath(file_path)
+
+    except ValueError:
+        pass
+
+    base, suffix = os.path.splitext(file_path)
+
```

Comment:
I'm not sure about the code that follows. The *kind* can be deduced from the type of loader (at least for loaders in the stdlib).

Reply:
I've replaced extension checking with `isinstance` checks against the concrete loaders in `importlib.machinery`. I agree that it appears clearer and more robust than `imp`-style suffix checks.